File: cbninterp.hs
---------------------------------------------------------------------------------------------------------
-- A monadic, higher-order, call-by-name interpreter based on [1] 
-- extended with case expressions.
--
-- [1] Monad Transformers Step by Step
-- [https://page.mi.fu-berlin.de/scravy/realworldhaskell/materialien/monad-transformers-step-by-step.pdf]
---------------------------------------------------------------------------------------------------------

module CBNMonadicInterp where

import Control.Monad.Except
import Control.Monad.Identity
import Control.Monad.Reader
import qualified Data.Map as Map
import Data.Maybe

type Name = String
 
type Env = Map.Map Name Expr

-- The result type of the eval function defined with monad transformers.
type Eval a = ReaderT Env (ExceptT String Identity) a

-- Binary operator constructors.
data OprName = Add | Subtr | Mult
             deriving Show

            
-- Expression representation
data Expr = IntExpr Integer
          | Var Name 
          | BinOp OprName (Expr, Expr)
          | Constr Name [Expr]
          | Case Expr [Branch]
          | Abs Name Expr
          | App Expr Expr
          | Thunk Env Name Expr
          deriving Show


-- Branch type consisting of a pattern and an associated 
-- expression which is evaluated in case pattern matches.          
data Branch = Branch Pattern Expr 
            deriving Show


-- A pattern is an expression that is an integer literal (constant), a variable, or a constructor. 
data Pattern = IntExprp Integer
             | Varp Name
             | Constrp Name [Pattern]
             deriving Show


---------- Helper functions ------------

-- Binary operators.
binOp :: OprName -> Expr -> Expr -> Integer
binOp opr (IntExpr x) (IntExpr y) = 
    case opr of
        Add   -> x + y
        Subtr -> x - y
        Mult  -> x * y


-- Insert a list of pairs (bindings) into an existing env.
updateEnv :: Env -> [(String, Expr)] -> Env
updateEnv env [] = env
updateEnv env (x:xs) = updateEnv (Map.insert (fst x) (snd x) env) xs


-- Take a list of variables and store their IDs in a list 
extractFromList :: [Pattern] -> [String]
extractFromList (Varp x:exs) = x : extractFromList exs
extractFromList [] = []


-- Evaluate the expression from the first branch whose pattern matches the 
-- case argument expression. If such a branch does not exist throw an error.
selectBranch :: Expr -> [Branch] -> Eval Expr
selectBranch c1@(Constr cname clist) (br:brs) = do
    env <- ask
    case br of
        Branch (IntExprp _) _ -> selectBranch c1 brs
        Branch (Constrp pcname plist) expr 
            | cname == pcname -> 
                let varlist = zip (extractFromList plist) clist
                    env' = const $ updateEnv env varlist
                in local env' (eval expr)
            | otherwise       -> 
                selectBranch c1 brs
        Branch p@(Varp vname) expr -> 
            local (const (updateEnv env [(vname, c1)])) (eval expr)
selectBranch c1@(IntExpr n) (br:brs) = do
    env <- ask
    case br of
        Branch (IntExprp p) expr 
            | n == p -> eval expr
            | otherwise -> selectBranch c1 brs
        Branch (Varp p) expr ->
            let env' = const $ updateEnv env [(p, c1)]
            in local env' (eval expr) 
        _ -> selectBranch c1 brs
selectBranch v (br:brs) = selectBranch v brs
selectBranch v [] = error "Unmatched"


-------------------- Eval function ---------------------------
--------------------------------------------------------------
eval :: Expr -> Eval Expr
-- Integer expression
eval (IntExpr n) = return $ IntExpr n
-- Variables now hold expressions, not just values. 
-- Thus, always evaluate the content of a variable.
eval (Var n) = do
    env <- ask
    case Map.lookup n env of
        Nothing -> throwError $ "Unbound Variable" ++ n
        Just val -> eval val
-- Binary operators
eval (BinOp operator (x, y)) = do
    e1 <- eval x
    e2 <- eval y
    case (e1, e2) of
        (IntExpr i, IntExpr j) ->         
            return $ IntExpr (binOp operator e1 e2)
        _ -> throwError "Invalid operands in binop"
-- Constructors. Do not change anything. 
eval (Constr cname exlist) = return $ Constr cname exlist
-- Abstractions. Return a thunk.
eval (Abs n e) = do
    env <- ask
    return $ Thunk env n e
-- Function application.
eval (App e1 e2) = do 
    val1 <- eval e1
    case val1 of
        Thunk env' n fbody -> 
            local (const (Map.insert n e2 env')) (eval fbody)
        _ -> throwError "Invalid function application"
-- Case expressions.
eval (Case expr brlist) = do
    evexpr' <- eval expr
    selectBranch evexpr' brlist


runEval :: Env -> Eval a -> Either String a
runEval env ev = runIdentity (runExceptT (runReaderT ev env))


----------- Test Cases -----------

-- (λx.λy.x + y) 2 10
test = runEval Map.empty
    (eval 
        (App
            (App        
                (Abs "x" 
                    (Abs "y" 
                        (BinOp Add (Var "x", Var "y")))) (IntExpr 2)) (IntExpr 10)))


-- x + y                        
test1 = runEval (Map.fromList [("x", IntExpr 1), ("y", IntExpr 2)])
    (eval
        (BinOp Add (Var "x", Var "y")))


-- λx.x+3
test2 = runEval (Map.fromList [])
    (eval  
        (Abs "x" 
            (BinOp Add (Var "x", (IntExpr 3)))))


-- (λx.λy.x+y) 2 bot
test3 = runEval (Map.fromList [])
    (eval 
        (App      
            (App
                (Abs "x" 
                    (Abs "y" 
                        (BinOp Add (Var "x", Var "y")))) (IntExpr 2)) (undefined)))


-- (λx.λy.x+1) 2 bot
test4 = runEval (Map.fromList [])
    (eval
        (App      
            (App
                (Abs "x" 
                    (Abs "y" 
                        (BinOp Add (Var "x", IntExpr 1)))) (IntExpr 2)) (undefined)))


-- case λx.x 5 of 
--   5          -> 1 + 1
--   List(x, y) -> x
test5 = runEval (Map.fromList []) 
    (eval  
        (Case (App (Abs "x" (Var "x")) (IntExpr 5)) [
            Branch (IntExprp 5) (BinOp Add (IntExpr 1, IntExpr 1)),
            Branch (Constrp "List" [Varp "x", Varp "y"]) (Var "x")]
        )
    )


-- case List(λx.x 5) of
--   List(y) -> y
--   5       -> 1 + 1
test6 = runEval (Map.fromList [])
    (eval 
        (Case (Constr "List" [(App (Abs "x" (Var "x")) (IntExpr 5))]) [
            Branch (Constrp "List" [Varp "y"]) (Var "y"),
            Branch (IntExprp 5) (BinOp Add (IntExpr 1, IntExpr 1))]
        )
    )


-- case List(λx.x bot) of
--   List(y) -> 1
--   5       -> 1 + 1
test7 = runEval (Map.fromList []) 
    (eval  
        (Case (Constr "List" [(App (Abs "x" (Var "x")) (undefined))]) [
            Branch (Constrp "List" [Varp "y"]) (IntExpr 1),
            Branch (IntExprp 5) (BinOp Add (IntExpr 1, IntExpr 1))]
        )
    )


-- case List(bot) of
--   List(y) -> 2
--   5       -> 
test8 = runEval (Map.fromList [])
    (eval 
        (Case (Constr "List" [(undefined)]) [
            Branch (Constrp "List" [Varp "y"]) (IntExpr 2),
            Branch (IntExprp 5) (BinOp Add (IntExpr 1, IntExpr 1))]
        )
    )


-- case List(λx.x 5) of
--   List(y) -> y
--   5       -> 1 + 1
test9 = runEval (Map.fromList [])
    (eval 
        (Case (Constr "List" [(App (Abs "x" (Var "x")) (IntExpr 5))]) [
            Branch (Constrp "List" [Varp "y"]) (Var "y"),
            Branch (IntExprp 5) (BinOp Add (IntExpr 1, IntExpr 1))]
        )
    )


-- case List(λx.x 5) of
--   List(y) -> d + 4
--   5       -> 1 + 1
test10 = runEval (Map.fromList [("d", IntExpr 3)])
    (eval 
        (Case (Constr "List" [(App (Abs "x" (Var "x")) (IntExpr 5))]) [
            Branch (Constrp "List" [Varp "y"]) (BinOp Add (Var "d", IntExpr 4)),
            Branch (IntExprp 5) (BinOp Add (IntExpr 1, IntExpr 1))]
        )
    )


-- 4 + ((λx.x + 3) 2)
test11 = runEval (Map.fromList [])
    (eval 
        (BinOp Add (IntExpr 4, 
            App
                (Abs "x" 
                    (BinOp Add (Var "x", (IntExpr 3)))) (IntExpr 2)
            )
        )
    )


-- case List(λx.x 5) of
--   List(y) -> λz.z 0
--   5       -> 1 + 1           
test12 = runEval (Map.fromList [("d", IntExpr 3)])
    (eval 
        (Case (Constr "List" [(App (Abs "x" (Var "x")) (IntExpr 5))]) [
            Branch (Constrp "List" [Varp "y"]) (App (Abs "z" (Var "z")) (IntExpr 0)),
            Branch (IntExprp 5) (BinOp Add (IntExpr 1, IntExpr 1))]
        )
    )

File: OCaml/cbninterp.ml
(*
 * An abstract machine, i.e., a call-by-name interpreter for the defined language.
 *)

type name = string

type operator = Add | Subtr | Mult

(*
 * This is the pattern of a case expression, i.e., the part
 * of the branch which case's argument is matched against.
 *)
type pattern = IntExprp of int
             | Varp of name
             | Constrp of name * (pattern list)

(*
 * Language's terms (expressions).
 *)
type expr = IntExpr of int                          (* Integer expression *)
          | Var of (name)                           (* Variable *)
          | BinOp of (operator * (expr * expr))     (* Binary operation *)
          | Constructor  of (name * (expr list))    (* Constructor, e.g., (cons (x nil)) *)
          | Abs of (name * expr)                    (* Lambda abstraction *)
          | App of (expr * expr)                    (* Function application *)
          | Case of (expr * (branch list))          (* Case expression *)


and  branch = Branch of (pattern * expr)

(*
 * A value is an integer, a constructor, or a thunk.
 *)
type value = IntVal of int
           | Thunk of (env * name * expr)
           | ConstrVal of (name * (value list))

(*
 * The environment, where values are stored.
 *)
and env = Env of ((name * value) list)


(* val : extractFromList : pattern list -> string list *)
let rec extractFromList xs = match xs with
    | (Varp x::xs) -> x :: extractFromList xs
    | _ -> []

(* 
 * Take an environment and a list of expressions and store each one 
 * in its own thunk.
 *)
(* val listOfThunks : env -> expr list -> value list *)
let rec listOfThunks env exprlist = match exprlist with
    | (x::xs) -> Thunk ((Env env), "_", x) :: listOfThunks env xs
    | []      -> []


(*
 * Evaluation function.
 *)
(* val eval : env -> expr -> value *)
let rec eval env = function
    | IntExpr n -> IntVal n
    (* 
     * A variable is bound to either an integer value or a thunk (i.e., a λ_.e).
     * In the first case, just return the value while in the second
     * return the evaluated expression of the thunk.
     *)
    | Var x -> 
        begin match List.assoc x env with
            (* 
             * If x is either an integer or a constructor, just return it. Otherwise, 
             * if it's a thunk, evaluate its body.
             *)
            | IntVal n as intv -> intv
            | ConstrVal (name, explist) as const -> const
            | Thunk (Env env, str, expr) -> eval env expr
        end
    (* The evaluation of an abstraction is a thunk. *)
    | Abs (n, ex) -> Thunk ((Env env), n, ex)
    (* Function application. *)
    | App (ex1, ex2) ->
        begin match eval env ex1 with
            (* Make sure ex1 evaluates to a thunk... *)
            | Thunk (Env en, n, ex) ->
                (* ...and build a second one with ex2. *)
                let thunk = Thunk ((Env env), "_", ex2)
                (* 
                 * Finally, insert the second thunk into the environment and evaluate ex1. 
                 * If during ex1's evaluation ex2's value is needed, then (and only then) evaluate it.
                 *)
                in eval ((n, thunk) :: en) ex
            | _ -> failwith "Invalid function application"
        end
    (* Constructor evaluation. *)
    | Constructor (cname, explist) ->
        (* Return the constructor with its arguments evaluated. *)
        let vallist = listEval env explist
        in ConstrVal (cname, vallist)
    (* Case expression evaluation. *)
    | Case (ex, (brlist)) ->
        matchPattern env ex brlist
    (* Binary operation evaluation. *)
    | BinOp (operator, (op1, op2)) -> 
        let x = eval env op1 in
        let y = eval env op2 in
            begin match (x, y) with 
                (IntVal o1, IntVal o2) -> IntVal (binOp operator o1 o2)
                | _ -> failwith "Invalid binary operands"
            end


(* val : listEval -> env -> expr list -> value list *)
and listEval env exprlist =
    match exprlist with
        | (x::xs) -> eval env x :: listEval env xs
        | []      -> []


(* val binOp : operator -> int -> int -> int *)
and binOp oprtr op1 op2 = match oprtr with
    | Add   -> op1 + op2
    | Subtr -> op1 - op2
    | Mult  -> op1 * op2


(*
 * matchPattern evaluates the expression part of the branch whose pattern matches the case argument.
 *
 *) 
(* val matchPattern : env -> expr -> branch list -> value *)
and matchPattern env casexp blist = match blist with
    | br::brlist ->
        begin match br with
            (* Case analysis on the patterns. *)
            | Branch (IntExprp n, pexp) ->
                (* If pattern is an integer value then we're forced to evaluate case's argument eagerly. *)
                let evex = eval env casexp in
                begin match evex with
                    | IntVal m -> 
                        if n = m then
                            (* If match succeeded, evaluate the associated expression. *)
                            eval env pexp
                        else
                            (* Otherwise, move to the next branch. *)
                            matchPattern env casexp brlist
                    (* Likewise, if the evaluated expression isn't an integer, then move to the next branch. *)
                    | _ -> matchPattern env casexp brlist
                end
            (* Pattern is constructor. *)
            | Branch (Constrp (pname, plist), cexpr) ->
                begin match casexp with
                    (* 
                     * If case's argument is a constructor and its name equals that of the pattern,
                     * insert its arguments in a list of thunks, append the list to env 
                     * and evaluate the associated expression.
                     *)
                    | Constructor (cname, clist) ->
                        if pname = cname then
                            let thunklist = listOfThunks env clist in
                            let env' = List.combine (extractFromList plist) thunklist in
                            eval (List.append env env') cexpr
                        else
                            matchPattern env casexp brlist
                    (*
                     * If it's an application, then make its argument a thunk
                     * and call matchPattern again with the body of the abstraction.
                     * This way, we bring the first argument to weak head normal form.
                     *)
                    | App (ex1, ex2) -> 
                        begin match  ex1 with
                            | Abs (str, expr1) -> 
                                let expr2 = (str, Thunk (Env [], "_", ex2)) in
                                matchPattern (expr2::env) expr1 blist
                            | _ -> failwith "Invalid case expression"
                        end
                    | _ -> matchPattern env casexp brlist
                end
            (* Pattern is variable. Matching is irrefutable. *)
            | Branch (Varp x, vexp) ->
                let thunk = Thunk ((Env env), "_", casexp) in 
                let env' = (x, thunk) :: env in
                    eval env' vexp
        end
    | [] -> failwith "Unmatched"


(* Test Cases (TODO: They should be in a separate file) *)

(* 4 + 2 *)
let test1 = 
    eval [] (BinOp (Add, (IntExpr 4, IntExpr 2)));;

(* λx.1 + 2 *)
let test2 = 
    eval [] (Abs ("x", (BinOp (Add, (IntExpr 1, IntExpr 2)))));;

(* λx.x + 2 *)
let test3 = 
    eval [("x", IntVal 2)] (Abs ("x", (BinOp (Add, (Var "x", IntExpr 2)))));;

(* (λx.x + 2) 10 *)
let test4 = 
    eval [("y", IntVal 2)] 
        (App  (Abs ("x", BinOp (Add, (Var "x", IntExpr 2))), IntExpr 10));;


(*
 * Omega, DOES NOT HALT. (λx.x x) (λx.x x)
 *
 *let test5 = 
 *  eval []
 *       (App (Abs ("x", App (Var "x", Var "x")), Abs ("x", App (Var "x", Var "x"))));;
 *)

(*
 * case (λx.x) 5 of
 *   5          -> 1 + 1
 *   List(x, y) -> x
 *)
let test6 = 
    eval []
        (Case (App (Abs ("x", Var "x"), (IntExpr 5)), [
            Branch ((IntExprp 5), (BinOp (Add, (IntExpr 1, IntExpr 1))));
            Branch (Constrp ("List", [Varp "x"; Varp "y"]), (Var "x"))])
        );;

(*
 * case List((λx.x) 5) of
 *   List(y)    -> y
 *   5          -> 1 + 1
 *)
let test7 = 
    eval []
        (Case (Constructor ("List", [App (Abs ("x", (Var "x")), (IntExpr 5))]), [
            Branch (Constrp ("List", [Varp "y"]), (Var "y"));
            Branch (IntExprp 5, BinOp (Add, (IntExpr 1, IntExpr 1)))])
        );;

(* 
 * case Cons1 omega of
 *   Cons1 y -> 2
 *   Cons2 y -> 1 + 2
 *)
let test8 = 
    eval []
        (Case (Constructor 
                ("Cons1", [((App ( 
                    Abs ("x", App (Var "x", Var "x")), 
                    Abs ("x", App (Var "x", Var "x")))))]), [
            Branch (Constrp ("Cons1", [Varp "y"]), (IntExpr 2));
            Branch (Constrp ("Cons2", [Varp "y"]), BinOp (Add, (IntExpr 1, IntExpr 2)))])
        );;

(* 
 * DOES NOT HALT!
 *
 * case List omega of
 *   List y -> y
 *   5      -> 1 + 1
 *)
(*
let test9 = 
    eval []
        (Case 
            (Constructor ("List", [((App 
                (Abs ("x", App (Var "x", Var "x")), 
                 Abs ("x", App (Var "x", Var "x")))))]), 
           [Branch (Constrp ("List", [Varp "y"]), (Var "y"));
            Branch (IntExprp 5, BinOp (Add, (IntExpr 1, IntExpr 1)))])
        );;
*)

let test10 = 
    eval []
        (Case (App (Abs ("x", (Constructor ("Cons1", [Var "x"]))), (IntExpr 51)), [
            Branch (Constrp ("Cons1", [Varp "y"]), (IntExpr 100))])
        );;
            

(* 
 * case (λx.Cons1[x]) Ω of
 *   Cons1[y] -> y
 *)
let test11 = 
    eval []
        (Case (App (Abs ("x", (Constructor ("Cons1", [Var "x"]))), (App 
                (Abs ("x", App (Var "x", Var "x")), 
                 Abs ("x", App (Var "x", Var "x"))))), [
            Branch (Constrp ("Cons1", [Varp "y"]), (IntExpr 1))])
        );;
